File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        long maxp = nums[0];
        long minp = nums[0];
        long res = nums[0];

        for(int i = 1 ; i<nums.length ; i++){
            int x = nums[i];

            if(x<0){
                long temp = maxp;
                maxp=minp;
                minp=temp;
            }

            maxp = Math.max(x,maxp*x);
            minp = Math.min(x,minp*x);

            res = Math.max(maxp,res);
        }
        return (int)res;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
    int maxProd = nums[0];
    int minProd = nums[0];
    int result = nums[0];

    for (int i = 1; i < n; ++i) {
        // If current number is negative, swap max and min
        if (nums[i] < 0) {
            swap(maxProd, minProd);
        }
        maxProd = max(nums[i], maxProd * nums[i]);
        minProd = min(nums[i], minProd * nums[i]);

        result = max(result, maxProd);
    }
    return result;
    }
};